### PR TITLE
Expose current NTP time along with associated analytical metadata

### PR DESCRIPTION
**Pull request by ryanrhee**
Time since last NTP sync can be a good indicator of the "fresh"-ness of a vended NTP time.

Does not replace existing APIs; callers can continue to get the timestamp as a `TimeInterval` directly rather than a struct that wraps it.

### CODE REVIEW

File: Sources/TimeFreeze.swift [hunk 1]
```diff
@@ -20,10 +21,23 @@ struct TimeFreeze {
         return (TimeFreeze.systemUptime() - self.uptime) + self.timestamp
     }
 
+    /// Returns the time plus analytical metadata about the time.
+    ///
+    /// See explanation of fields in the comment for the FullTime struct fore more info.
+    var fullTime: FullTime? {
```
Comment (Reflejo) [inline Sources/TimeFreeze.swift]: I'd expose a `var timeSinceLastNTPSync` as opposed to this `FullTime` new struct

File: Sources/Clock.swift [hunk 2]
```diff
@@ -32,7 +44,11 @@ public struct Clock {
 
     /// The most accurate date that we have so far (nil if no synchronization was done yet)
     public static var now: Date? {
-        return self.timestamp.map { Date(timeIntervalSince1970: $0) }
+        return self.fullNow?.date
+    }
+
+    public static var fullNow: FullTime? {
```
Comment (Reflejo) [inline Sources/Clock.swift]: missing docstring, and same comment as below, I'd expose this as `timeSinceLastNTPSync`

File: Sources/TimeFreeze.swift [hunk 3]
```diff
@@ -40,9 +54,11 @@ struct TimeFreeze {
         if rint(currentBoot) - rint(previousBoot) != 0 {
             self.uptime = currentUptime
             self.timestamp = currentTimestamp
+            self.isAdjustedForReboot = true
```
Comment (Reflejo) [inline Sources/TimeFreeze.swift]: I'd store the NTP sync time on a state as well as into disk using the same mechanism we use for storing times, right after sync'ing (and update it every time we hit an NTP server). I'd get rid of this flag; this way on a reboot we won't restore that date (reminding nil for the period of time while we are sync'ing the clock)

**Comment by ryanrhee** [comment]
Having separate getters for time since last sync and timestamp is
inherently prone to race conditions. This exact same conversation happened
in Android as well and we ended up with this kind of struct :)

I'm not sure I follow your suggestion on storing NTP sync time on disk.
Could you elaborate?

**Comment by Reflejo** [comment]
> Having separate getters for time since last sync and timestamp is inherently prone to race conditions.

Can you explain how is it inherently prone to race conditions? Your computed property is doing the math the same way, where are you thinking the race condition will be exposed?

**Comment by ryanrhee** [comment]
Yes, happy to explain!

So if clock has a property for time since last sync and current ntp time,
you can imagine consumers to do something like this:

time = clock.currentTime
age = clock.timeSinceLastSync

In this scenario, if there's a repeating clock sync that happens every N
minutes, it's possible that NTP sync happens between the two lines that
execute.

In that case, the timestamp's actual age would be high, while the actual
age returned would be near 0. (The reverse could happen if you swap the two
lines.)

Hope that makes sense! I'm happy to explain in more detail if necessary.

**Comment by Reflejo** [comment]
@ryanrhee Easier to explain with code what I meant: https://github.com/lyft/Kronos/compare/fz-sync-time?expand=1

**Comment by ryanrhee** [comment]
OK I understand what you mean now, thanks for explaining.

I'm worried to take this approach because we're relying on the system clock to provide us a datetime for us to calculate the age of the NTP response, as return values of `Date()` can change between reboots. To avoid this, I think we'd have to use a stable clock, but because our stable clock measures time since boot, we wouldn't be able to correlate across reboots.

**Comment by Reflejo** [comment]
@ryanrhee Ah, ok. I didn't realized you needed this annotation to be so trustful. That's fine. You'll still be able to simplify this and remove the boolean after https://github.com/lyft/Kronos/pull/50/files

re tuple: Yea I'd rather have a tuple, which IMO is a semantic that explains a bit better the relationship between two properties that not necessarily form an identity.

**Comment by ryanrhee** [comment]
Generally I'm of the opinion that the annotation should always be trustful, whereas the actual NTP time can be more or less trustful depending on needs. This is because annotations are used during investigations, and having not-trustful metrics makes investigation a lot more difficult.

**Comment by ryanrhee** [comment]
And OK, will change to a Tuple.

File: Sources/TimeFreeze.swift (hunk 1 above)
Comment (ryanrhee) [reply Sources/TimeFreeze.swift]: can't do this bc it's racy, explanation below

File: Sources/Clock.swift (hunk 2 above)
Comment (ryanrhee) [reply Sources/Clock.swift]: can't do this bc it's racy, explanation below

File: Sources/TimeFreeze.swift (hunk 3 above)
Comment (ryanrhee) [reply Sources/TimeFreeze.swift]: doing so would make the age less trustful, which we want to avoid. discussion below.

**Comment by Reflejo** [comment]
Sounds good, once you rebase you'll be able to remove the bool

File: Sources/Clock.swift [hunk 4]
```diff
@@ -1,5 +1,8 @@
 import Foundation
 
+/// Struct that has time + related metadata
+public typealias FullTime = (Date, TimeInterval)
```
Comment (Reflejo) [inline Sources/Clock.swift]: please use named arguments, i.e.

```
public typealias FullTime = (date: Date, timeSinceLastNtpSync: TimeInterval)
```

File: Sources/TimeFreeze.swift [hunk 5]
```diff
@@ -20,6 +20,14 @@ struct TimeFreeze {
         return (TimeFreeze.systemUptime() - self.uptime) + self.timestamp
     }
 
+    /// Returns the time plus analytical metadata about the time.
+    ///
+    /// See explanation of fields in the comment for the FullTime struct fore more info.
+    var fullTime: FullTime? {
```
Comment (Reflejo) [inline Sources/TimeFreeze.swift]: shouldn't be an optional anymore

File: Sources/Clock.swift [hunk 6]
```diff
@@ -32,7 +35,11 @@ public struct Clock {
 
     /// The most accurate date that we have so far (nil if no synchronization was done yet)
     public static var now: Date? {
-        return self.timestamp.map { Date(timeIntervalSince1970: $0) }
+        return self.fullNow?.0
+    }
+
+    public static var fullNow: FullTime? {
```
Comment (Reflejo) [inline Sources/Clock.swift]: missing docstring

File: Sources/TimeFreeze.swift [hunk 7]
```diff
@@ -20,6 +20,14 @@ struct TimeFreeze {
...
+        return FullTime(date: Date(timeIntervalSince1970: adjustedTimestamp),
+                        timeSinceLastNtpSync: TimeFreeze.systemUptime() - uptime)
```
Comment (Reflejo) [inline Sources/TimeFreeze.swift]: This is optional: I prefer for `TimeFreeze` to know nothing about composing this, but expose the `timeSinceLastNtpSync` instead. So you'd have `var timeSinceLastNtpSync { ... }` here and compose this `FullTime` thing on the `Clock` interface by doing something like:

```swift
public static var fullTime: FullTime? {
    guard let stableTime = self.stableTime else {
        return nil
    }

    return FullTime(date: Date(timeIntervalSince1970: stableTime.adjustedTimestamp),
                    timeSinceLastNtpSync: TimeFreeze.systemUptime() - stableTime.uptime)
}
```

File: Sources/TimeFreeze.swift [hunk 8]
```diff
@@ -20,6 +20,14 @@ struct TimeFreeze {
         return (TimeFreeze.systemUptime() - self.uptime) + self.timestamp
     }
 
+    /// Returns the time plus analytical metadata about the time.
+    ///
+    /// See explanation of fields in the comment for the FullTime struct fore more info.
+    var fullTime: FullTime? {
+        return FullTime(date: Date(timeIntervalSince1970: adjustedTimestamp),
```
Comment (Reflejo) [inline Sources/TimeFreeze.swift]: nit: missing `self.` on both this and the next line.

Comment (ryanrhee) [reply Sources/TimeFreeze.swift]: n/a since time freeze no longer knows how to compose this

File: Sources/TimeFreeze.swift (hunk 7 above)
Comment (ryanrhee) [reply Sources/TimeFreeze.swift]: done!

File: Sources/Clock.swift (hunk 6 above)
Comment (ryanrhee) [reply Sources/Clock.swift]: added

File: Sources/TimeFreeze.swift (hunk 5 above)
Comment (ryanrhee) [reply Sources/TimeFreeze.swift]: no longer optional

File: Sources/Clock.swift (hunk 4 above)
Comment (ryanrhee) [reply Sources/Clock.swift]: added named arguments, and changed name to "AnnotatedTime"

File: Tests/KronosTests/NTPClientTests.swift [hunk 9]
```diff
@@ -40,7 +40,7 @@ final class NTPClientTests: XCTestCase {
 
     func testQueryPoolWithIPv6() {
         let expectation = self.expectation(description: "NTPClient queries a pool that supports IPv6")
-        NTPClient().query(pool: "2.pool.ntp.org", numberOfSamples: 1) { offset, _, _ in
+        NTPClient().query(pool: "2.pool.ntp.org", numberOfSamples: 1, maximumServers: 1) { offset, _, _ in
```
Comment (ryanrhee) [inline Tests/KronosTests/NTPClientTests.swift]: tests passing without `maximumServers` being set was a coincidence, and this test change is unrelated to the code AFAICT.

File: Sources/Clock.swift [hunk 10]
```diff
@@ -32,7 +42,17 @@ public struct Clock {
 
     /// The most accurate date that we have so far (nil if no synchronization was done yet)
     public static var now: Date? {
-        return self.timestamp.map { Date(timeIntervalSince1970: $0) }
+        return self.annotatedlNow?.0
```
Comment (Reflejo) [inline Sources/Clock.swift]: I guess you don't need this change now? alternatively use `.data` as opposed to `.0`

File: Tests/KronosTests/NTPClientTests.swift (hunk 9 above)
Comment (Reflejo) [reply Tests/KronosTests/NTPClientTests.swift]: thanks for fixing it!

File: Sources/Clock.swift (hunk 10 above)
Comment (ryanrhee) [reply Sources/Clock.swift]: good catch, thx

**Comment by ryanrhee** [comment]
retrying CI

File: Sources/Clock.swift [hunk 11]
```diff
@@ -32,7 +42,17 @@ public struct Clock {
 
     /// The most accurate date that we have so far (nil if no synchronization was done yet)
     public static var now: Date? {
-        return self.timestamp.map { Date(timeIntervalSince1970: $0) }
+        return self.annotatedlNow?.date
+    }
+
+    /// Same as `now` except with analytic metadata about the time
+    public static var annotatedlNow: AnnotatedTime? {
```
Comment (Reflejo) [inline Sources/Clock.swift]: Is this a typo? `annotated<l>Now`?
